### ISSUE TITLE
CLOGO document

### ISSUE THREAD

**Issue by Alexey-Slyusar** [PDP-10/its#1197]
Here is a link to a relatively comprehensive description of the CLOGO Language and System
This is a very early draft of the document and a lot are missing. I am working to get it finished.

[A Description of the LOGO Language and System](https://drive.google.com/file/d/1M7lHjbgk1sM7v87cwjrZ2jw2iijVqfSI/view?usp=sharing)

I'm going to make a new branch alexey/CLOGO to add the document.
@larsbrinkhoff, could you please suggest a name for the document. Is "doc/_info_/clogo.order" name in line with the ITS practice?

**Comment by larsbrinkhoff** [comment]
Great, thanks!

The files named ORDER are usually brief lists of commands.  Your document is more comprehensive.  I'd say .INFO.; CLOGO MANUAL or similar would be in order.  In the Unix file system, that would be doc/_info_/clogo.manual.

**Comment by Alexey-Slyusar** [comment]
I will follow the convention :-)
I've made the alexey/CLOGO branch and put draft to doc/_info_/clogo.manual.
Thank you!

**Comment by Alexey-Slyusar** [comment]
Hi, Lars, 

I am working on part 2.5 "Filing". There are SAVE and GET procedures in the CLOGO vocabulary that serve, respectively, to save and retrieve an entry from a file. As far as I am understand, the procedures accept two argument: directory name and file name. I did some try and error iterations, but I still cannot figure out how to provide right form of the arguments to the procedures.

Here is short session:
?SAVE
TWO FILE NAMES NEEDED.
?SAVE "ALEXEY;" "LOGO"
THERE IS NO FILE DIRECTORY.
?SAVE "HQM;" "LOGO"
THERE IS NO FILE DIRECTORY.

Do you have any ideas how to provide right form of the arguments to the procedures?
Thanks in advance!

**Comment by Alexey-Slyusar** [comment]
@eswenson1, thank you!

I've tried things out, but unsuccessfully.
Here is result: 

:CWD .TEMP.;
*:СLOGO

LOGO.44    
?SAVE FOO BAR
THERE IS NO FILE DIRECTORY.
?SAVE "FOO" "BAR"
THERE IS NO FILE DIRECTORY.
?SAVE ".TEMP." "FOO.BAR"
THERE IS NO FILE DIRECTORY.
?SAVE ".TEMP.;" "FOO.BAR"
THERE IS NO FILE DIRECTORY.
?GB

AND A PLEASANT DAY TO YOU!   
:listf

DB   .TEMP.
FREE BLOCKS #0=19739

I'll do more exploration in the evening.

Thanks again!

**Comment by eswenson1** [comment]
Maybe the directory name is hard coded, but doesn’t exist? Where is the source for CLOGO?

-- Eric

**Comment by larsbrinkhoff** [comment]
We have no source code.  And not even a clue where the program comes from; whether it's a port or originally by MIT.

**Comment by Alexey-Slyusar** [comment]
The sintax and general behaviour of the  CLOGO are almost identical with
BBN LOGO. Thus, I think CLOGO is probably the MIT port of the BBN LOGO and
the earliest MIT Logo System implementation.
But it is only my humble opinion.

P.S. There  are no the DISPLAY TURTLE related procedures in CLOGO, only
floor turtle routines are presented. So, in any case, it is an earlier
implementation than MIT 11LOGO and MIT LISP LOGO.
But we have very stable CLOGO binary version, circa 1975 or so if I am not
mistaken.

чт, 13 сент. 2018 г. в 9:01, Lars Brinkhoff <notifications@github.com>:

**Comment by larsbrinkhoff** [comment]
I compared CLOGO disassembly with some kind of "BBN Logo", but they were not similar at all.

Our current version of CLOGO is from 1975.  There are also files going back to 1972.

**Comment by Alexey-Slyusar** [comment]
My humble opinion about the similarity between CLOGO and BBN LOGO is based
on the BBN reports which descibe the BBN LOGO language and system and my
own experience with our current version  of CLOGO.
Perhaps CLOGO is not a direct port, but an implementation written from
scratch that functionally repeats the BBN LOGO. Therefore they are
disassembly different but functionally and syntactically similar. But this
is only my guess :-)



чт, 13 сент. 2018 г. в 11:10, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
I'm going to try out whether the directory name is hard coded. Maybe I will be lucky enough to guess the name of the hard coded directory.To do that I going to create directories name like "LOGO", "CLOGO" etc. and try to save a file in.
Lars, you said that there are also files going back to 1972. Are these files stored in directories? If so, what are the names of the directories? 
I learned that the command for creating directory in ITS is :print dir;..new. (udir), but I cannot figure out how to use it properly. Could you please provide the exact command syntax for creating, for example, DSK:LOGO; directory? Thank you very much for your help and sorry for such silly question.

**Comment by eswenson1** [comment]
If you want to create a directory called FOO, do this:

:print foo;..new. (udir)<enter>

or:

^R foo; ..new. (udir)<enter>

**Comment by eswenson1** [comment]
You will get an error message when you do this.  Just ignore the error message.  Type FOO^F to see if the (empty) directory got created.  Note: if you don't put anything in the directory, when the system is rebooted, the empty directory will get automatically deleted.

**Comment by eswenson1** [comment]
You could read the binary file into emacs (on ITS) and look for "interesting strings".  You might be able to spot any hardcoded directory names.  Alternatively, you could try to find where the file i/o is happening and use DDT to debug the file i/o calls.  Of course, trying to find where to put the breakpoint is difficult when you have no sources and when the binary is large.

**Comment by Alexey-Slyusar** [comment]
Eric, thank you very much!

**Comment by larsbrinkhoff** [comment]
The binaries were in SYS and SYSBIN, so that's no clue unfortunately.

**Comment by Alexey-Slyusar** [comment]
Thank you!

I am using DrRacket already in accompany with the wonderful Simply Scheme book by Dr. Brian Harvey. It is a prequel for SICP by Abelson and Sussman. To be honest I tried SCIP but it is to complicated for me and I plan to read it after I will absorb all Simply Scheme ideas.

But at the same time I am huge logo fan. I very like logo and I think that the idea to have Logo and Lisp environment simultaneously is just amazing. So Lisp Logo is very interesting for me. 

Thanks again, Eric! I am happy to join ITS community there are many interesting people here and interesting things to play with :-)

**Comment by larsbrinkhoff** [comment]
Eric, the binary file is HQM; CLOGO BIN, and it's purified to SYS; TS CLOGO.

**Comment by larsbrinkhoff** [comment]
It's strange that we have a set of binaries in the period 1972-1976, but no source.  Maybe it's just because the source wasn't called CLOGO?

**Comment by Alexey-Slyusar** [comment]
The most obvious guesses for the hard coded directory name did not work.
I loaded the binary file to emacs and looking for "interesting strings". There is no result yet.

I get back with the result in the evening and I want to quote a CLOGO message here:
"AND A PLEASANT DAY TO YOU!"

**Comment by larsbrinkhoff** [comment]
EMACS will only display 7-bit ASCII text.  The directory names are likely to be a ASCII subset stored as 6-bit characters.

I have a [disassembler that runs in Unix](http://github.com/larsbrinkhoff/dis10).  It will display most information inside a binary, including symbol table and possible 7-bit and 6-bit strings.

Running on HQM; CLOGO BIN, I see:

- It's a PDUMP binary, so it's already purified.
- The start instruction is a JUMPA.  MIDAS uses JRST.
- The symbol table is just one global section.

Some possible directory or file names:

- LOGO, contents at symbol FSNAME
- .LOGO., contents at symbol DSNAME
- IPG, contents at symbol LSNAME
- LESSON
- DRIBLE
- LOGO BUGS
- LOGO NEWS
- RJL

Interesting uses of .SUSET:

- 5357/ .SUSET [.SSNAM,,FSNAME]
- 5363/ .SUSET [.SSNAM,,DSNAME]
- 5417/ .SUSET [.SSNAM,,FSNAME]
- 5445/ .SUSET [.SSNAM,,[SIXBIT /RJL/]]
- 5474/ .SUSET [.SSNAM,,[SIXBIT /LOGO/]]
- 13525/ .SUSET [.SSNAM,,FSNAME]
- 13604/ .SUSET [.SSNAM,,FSNAME]
- 14217/ .SUSET [.SSNAM,,NFRAME]
- 14537/ .SUSET [.SSNAM,,LSNAME]

**Comment by larsbrinkhoff** [comment]
From this, I tracked down some more information:

RJL = Ron Lebel, a major force in the MIT Logo group.

From https://logothings.github.io/logothings/The80s.html:

> Our first full year Logo class was in 1968-69 with seventh graders at Muzzey Junior High School. We were at BBN at the time. In the fall of 1969 I became a member of the AI Lab and we started our Logo work there.
> 
> The first graphics turtle and a floor turtle were built and running by
> 
> The graphics turtle was not portable. Its display was driven by a Digital PDP-6 connected to a Logo running on the lab’s PDP-10. First programs were written by Bill Gosper and other lab hackers. The yellow turtle was assembled by Tom Callahan.
> ![Yellow Turtle](https://user-images.githubusercontent.com/775050/45533108-d1882b00-b7f6-11e8-893f-dd7c67cf035a.jpg)

(Note the turtle lamp.)

Again looking inside CLOGO BIN, I see there's a PDP6O symbol pointing to ['USR ? 0 ? SIXBIT /PDP6/].  In other words, this program is using the attached PDP-6 computer.  So it looks like this is the version mentioned in the quote.

**Comment by larsbrinkhoff** [comment]
@natgoodman, we have found some more information about CLOGO.  If you look at the comment above this one, does it ring any bells?  Thank you!

**Comment by larsbrinkhoff** [comment]
IPG could be Ian P. Goulden.  Was he involved with Logo?

**Comment by natgoodman** [comment]
Hi Lars

I looked at the thread.  You’re making impressive progress!  The comment jogged my memory a tad.  I remember the Muzzy Junior High effort - not sure I ever went out there, it may have been just as I was joining the project.  I vaguely remember the yellow turtle.  By my time we had started the move to the little plexiglass dome covered beastie.  I don’t remember running the turtle from the PDP-6.  Again, perhaps before my time or just as I was joining the project.

I also checked out Cynthia Solomon’s website.  A trove of LOGO history!

All the best,
Nat

**Comment by eswenson1** [comment]
@Alexey-Slyusar I took Sussman's course back in 1976 and really like the SICP book.  And yes, my kids, who took a functional programming class in Scheme at their high school said that they used a book that followed SICP, but was modernized and geared to DrScheme use.  I suppose we shoudn't burden readers of this issue thread with off-topic stuff. Feel free to write me at eric at swenson.org if you want to chat further about this kind of stuff.  I'm currently a FP developer who uses Scala!

**Comment by Alexey-Slyusar** [comment]
The SAVE command doesn't accept the LOGO, .LOGO., LESSON directory names as an argument.
But there is interesting behavior:

:CWD .LOGO.;
*:LOGO

LOGO.44    
?TO HELLO
>10 PRINT "HELLO"
>END
HELLO DEFINED
?SAVE .LOGO. HELLO
THERE IS NO FILE DIRECTORY.
?SAVE ".LOGO." "HELLO"
THERE IS NO FILE DIRECTORY.
?SAVE .LOGO.; HELLO
ERROR AT 15357
? 

Please note that there is ERROR AT 15357 message appeared when SAVE evaluated with a semicolon at the end of the .LOGO. argument. Any ideas?

**Comment by Alexey-Slyusar** [comment]
It seems that not only .LOGO. actually, but in case of any argument with semicolon at the end:
SAVE FOO; BAR

**Comment by eswenson1** [comment]
And does it make any difference whether the directory exists or not?

**Comment by larsbrinkhoff** [comment]
I have found something that may be CLOGO sources in the LOGO directory.  But it'll be a while before we can see them.

**Comment by Alexey-Slyusar** [comment]
No, there is no difference whether the directory exist or not. The error is appeared in both cases.

**Comment by Alexey-Slyusar** [comment]
Oh, I just saw the message from Lars! 
So there is a hope to get CLOGO sources :-)

**Comment by Alexey-Slyusar** [comment]
In connection with the yellow turtle turtle lamp, there are LAMPON and LAMPOFF commands in the CLOGO vocabulary :-)

LOGO.44    
?LAMPON
YOU HAVE TO DO A SETTURTLE
?SETTURTLE T00
?LAMPON
ci"hb?
?LAMPOFF
ci!hb?

**Comment by larsbrinkhoff** [comment]
The terminal line is exposed outside the PDP-10 emulator as a TELNET port.  It would be possible to attach an external program which listens to the commands sent out to the line.  Maybe write it in a modern Logo! :-)

**Comment by Alexey-Slyusar** [comment]
Great idea, Lars, I must say :-)

To make a step further, there is Brian Silverman's LogoChip project.
The LogoChip is programmable and inexpensive embedded microcontroller (Microchip PIC18F2320 microcontroller that has LogoChip firmware installed). 
http://www.playfulinvention.com/portfolio/logo/?doing_wp_cron=1537000645.8897709846496582031250

Potentially, we can interface CLOGO with LogoChip and recreate the Yellow Turtle controlled by CLOGO :-)

By the way, what the number of the TELNET port that used by PDP-10 emulator?

P.S. Brian Silverman is a Consulting Scientist to the MIT Media Lab and notable Logo developer and cited by Brian Harvey as one of the contributors to the development of Logo at MIT.

**Comment by larsbrinkhoff** [comment]
The TELNET port is number 10000 with the KA10 simulator, and 10004 with SIMH.

I have tried to contact Brian Silverman a few times, but so far without success.

**Comment by Alexey-Slyusar** [comment]
Lars, thank you!
I think that your great idea to interface CLOGO with an external program deserve a separate issue, so I've created it. #1227
I hope you agree. :-)

**Comment by Alexey-Slyusar** [comment]
I tried the mentioned directories once again (except for LOGO BUGS and  LOGO NEWS*), but unsuccessful. 

I still don't figure out how SAVE works.

![default](https://user-images.githubusercontent.com/42380964/45772874-85ebdc00-bc51-11e8-97c3-1674ddae6f4b.png)


* I got LOGO B and LOGO N instead.

`^R LOGO.BUGS;..NEW. (UDIR)`
`DSK: LOGO.B; ..NEW. (UDIR) - FILE NOT FOUND`
`DSK: DIR LOGO.B CREATED BY  ALEXEY HACTRN  20:52:57`
`^R LOGO.NEWS;..NEW. (UDIR)`
`DSK: LOGO.N; ..NEW. (UDIR) - FILE NOT FOUND`
`DSK: DIR LOGO.N CREATED BY  ALEXEY HACTRN  20:53:29`

**Comment by larsbrinkhoff** [comment]
You did it right, but ITS directory names can only be six characters.

**Comment by Alexey-Slyusar** [comment]
AIM 307a says that proper syntax for `SAVE` command is:
`SAVE FOO BAR DSK HENRY`
We assume that the directory name is hard coded in CLOGO. But all of the attempts to find the proper name for the directory felled.
But maybe the disk name (I mean `DSK`) is hardcoded too? 
What is possible variants for the disk name?

**Comment by eswenson1** [comment]
Doesn’t “save foo bar dsk henry” imply that these four arguments are FN1, FN2, DEVICE, DIRECTORY? “DSK” is a standard device name, and the use of four components suggests a full ITS file name spec.

**Comment by larsbrinkhoff** [comment]
There are some new files.  I beleve these are much of the source code for CLOGO, though not complete enough to assemble them.  Maybe these can answer how SAVE is to be used.

https://github.com/PDP-10/its-vault/tree/master/files/rjl/

**Comment by Alexey-Slyusar** [comment]
Yes, Eric, it is imply four arguments exactly. But I thought that `DSK` may have different from the standard device name.

**Comment by Alexey-Slyusar** [comment]
Thank you for the link, Lars! I'll check the source files for possible `SAVE` usage hints.

**Comment by Alexey-Slyusar** [comment]
I've found the CLOGO error text that SAVE command produces:
```
?SAVE FOO BAR DSK LOGO
THERE IS NO FILE DIRECTORY.
```
in the `logo.49` source file (string no. 3685 of the file):

`EE NFLER1,THERE IS NO FILE DIRECTORY.,,`

and some SAVE related assembler code with `NFLER1:`
```
SAVE:	PUSHJ P,FILGET
	HRLZM E,TBL+1
	PUSHJ P,FILGET
	HRRM E,TBL+1
	MOVEI TA,TBL
	.SUSET [.SSNAM,,FSNAME]
	PUSHJ P,WRITEF
	JUMPE A,SAVE1
	CAIE A,NFLER3-NFLER1+1
	JRST FILERR
	MOVEI B,[ASCIZ \FILE ALREADY EXISTS.
DO YOU WANT TO DELETE IT?\]
	PUSHJ P,TOSS
	PUSHJ P,TYI
	MOVE E,C
	PUSHJ P,CRLF
	CAIE E,"Y
JRST RESET
```
But I still cannot figure out how exactly SAVE works.

Please note, that we have CLOGO version 44:
```
?PRINT VERSION  
44
```
So the `logo.49` source file is possible the source for a newer CLOGO version.

I examined the `logo.new` source file in the same directory also, but there is no `THERE IS NO FILE DIRECTORY` string in the file.

**Comment by Alexey-Slyusar** [comment]
Thanks for the comment, Lars!
It seems that intended for implement the same `SAVE` behaviour as we already discussed in #1230.
You wrote: 
> I tried SAVE FOO BAR, and it wrote a FOO LGO file in the DECSYS directory. The contents include 
> the text of a procedure I wrote.

I have the similar experience with SAVE except that a file stored in working directory. So if I run `BBN;LOGO` while being in `DSK:ALEXEY;` directory, the files will stored in `DSK:ALEXEY;`.

Please note, that due to the pedagogical reason, the SAVE command doesn't write file in convenient for ITS `FN1` `FN2` format, and uses ENTRY conception instead. 
So SAVE takes exactly two arguments. The first is for file name and the second for the entry name in a file, as it described in BBN Report No. 1889.:
![default](https://user-images.githubusercontent.com/42380964/49683027-40f87880-facf-11e8-8a62-f343d668e3d3.png)
And further there:
![default](https://user-images.githubusercontent.com/42380964/49683070-fd523e80-facf-11e8-9927-cee5ff35c14b.png)

It is very likely that CLOGO has a similar mechanism. 
So LISP LOGO `SAVE FOO BAR DSK HENRY` format not valid for CLOGO.

**Comment by Alexey-Slyusar** [comment]
I am still trying to figure out how SAVE command work in CLOGO. I am
exploring BBN's SAVE behavior - that works - to find any clue to solve the
pazzle. And here results:
If you logged in onto ITS as say LUSER but don't have LUSER'S directory on
ITS, then BBN's SAVE complains: "I AM IN TROUBLE.  TELL YOUR TEACHER". So,
BBN LOGO is login name sensitive as well as directory name. Possibly, CLOGO
is the same. And it is possible also that login name is hardcoded too
(teacher's login?), so all children in a LOGO class could login with it and
worked ony in the single special directory on ITS.
I am investigating in things further.


сб, 8 дек. 2018 г. в 13:25, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by larsbrinkhoff** [comment]
I put a breakpoint at SAVE and stepped forward.  First, it sets the directory to LOGO.  Second, in FILOPR, it opens a file called <uname> FILES.  So apparently there must be a LOGO; LARS FILES for this to work.  I don't know what it contains, or how it should be created the first time.

**Comment by larsbrinkhoff** [comment]
If I make it empty, CLOGO says

    FILE DIRECTORY FULL.

**Comment by Alexey-Slyusar** [comment]
It is big step further Lars, thank you very much!!!

I'll experiment with this new information and get back with results.

сб, 12 янв. 2019 г. в 7:45, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
I've got the first result.
Yes, I cannot SAVE and GET procedures to empty ALEXEY FILES file created
using, say, EMACS either. But I find some workaround way to get the file in
proper CLOGO format. I created ALEXEY LGO file with BBN LOGO with procedure
definition stored and rename it as ALEXEY FILES. As I wrote above in the
tread, BBN LOGO has some special file format. Luckily, BBN LOGO has CLOGO
compatible file format, so I can read stored procedure FOO as
?GET ALEXEY FOO
from CLOGO prompt.
But it seems that CLOGO SAVE command doesn't work properly with the file.
Despite It reports:
?SAVE ALEXEY BAR
45 CHARACTERS SAVED.
The BAR procedure haven't had stored in the file actually.


сб, 12 янв. 2019 г. в 8:05, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by Alexey-Slyusar** [comment]
It is interesting that my workaround approach doesn't work on ES. Possibly
because the path is ES:LOGO;ALEXEY FILES rather than DSK:LOGO;ALEXEY FILES
on DB.

сб, 12 янв. 2019 г. в 8:43, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by Alexey-Slyusar** [comment]
Thank you very much for explanation, Eric. So my guess was wrong.

Here my workaround approach on DB:
```
:login alexey

*:BBN;LOGO

LOGO

_TO FOO
>10 PRINT "BAR"
>END
FOO DEFINED
_SAVE ALEXEY FOO
_GB

:KILL 
*:COPY DSK:ALEXEY;ALEXEY LGO,DSK:LOGO;ALEXEY FILES
*:CWD LOGO
*:CLOGO

LOGO.44    
?TO FOO2
>10 PRINT "BAR2"
>END
FOO2 DEFINED
?SAVE ALEXEY FOO2
38 CHARACTERS SAVED.
```
Most possibly I did not exactly the same on ES.

Please note, that BBN LOGO / CLOGO files aren't in plain text format, but are in some kind of binary format.

**Comment by eswenson1** [comment]
I was able to reproduce your steps using both BBN LOGO and CLOGO on ES. I copied LOGO;ALEXEY FILES to LOGO; EJS FILES. Then both SAVEs worked fine.

-- Eric

**Comment by Alexey-Slyusar** [comment]
Eric, thanks for taking the time to look at this.
I :PRINT EJS FILES on ES and didn't see    FOO procedure definition in the
file. There is HELLO procedure definition only. Did you define HELLO or FOO
procedure?

пн, 14 янв. 2019 г. в 9:38, Eric Swenson <notifications@github.com>:

**Comment by eswenson1** [comment]
In my case, the files are being saved to EJS;  

So I saved to AAA AAA and the result was saved in EJS; AAA LGO.

When I saved to ALEXEY FILES, the result was saved in EJS; ALEXEY LGO.

I’m not sure why I have to specify two filenames, if the second is just ignored. 

If you look in my home directory, you should see EJS; AAA LGO and EJS; ALEXEY LGO and EJS; EJS LGO.

**Comment by Alexey-Slyusar** [comment]
It seems that SAVE doesn't working properly. It reports about succsess but
doesn't save a procedure definition in the file actually.

And despite that GET ALEXEY FOO doesn't complain it doesn't load FOO
procedure definition into CLOGO workspace either.

пн, 14 янв. 2019 г. в 10:01, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by Alexey-Slyusar** [comment]
But at the same time both versions have the same file format for store
workspace definitions. Our CLOGO version has a creating file problem as
Lars wrote above. That is why I use BBN LOGO to create ALEXEY FILES file
for CLOGO.

пн, 14 янв. 2019 г. в 10:24, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by larsbrinkhoff** [comment]
Maybe CLOGO has some other command to create a fresh empty directory
file.  I used DDT to scan back and forth from the SAVE label, but I
didn't see anything that looked like that.  It's hard when we don't
have the full source code, but maybe some day we will.

**Comment by Alexey-Slyusar** [comment]
Thank you, Lars!
One more source of difficulties here is that there is no CLOGO related MIT
MEMO at all. I use BBN REPORTS to explore CLOGO vocabulary and behavior.
One more possibly suitable CLOGO command for file storing is LESSON
command, but I don't know how it exactly works either.

Maybe "The LOGO Primer", LOGO Working Paper 19, MIT AI Lab (Jan. 1972) by
Austin, H. describes CLOGO dialect. But unfortunately, I cannot find the
document on the net.


пн, 14 янв. 2019 г. в 10:53, Lars Brinkhoff <notifications@github.com>:

**Comment by larsbrinkhoff** [comment]
I did ask some MIT people about CLOGO, but they don't even remember that
name.

So there isn't a complete set of Logo memos online?  I did get some help
from the MIT library services about AI memos.  You could try there.

https://libraries.mit.edu/forms/dspace-help/

https://libraries.mit.edu/forms/ask-the-archives-non-mit-community/

**Comment by Alexey-Slyusar** [comment]
Lars, thank you for the links!

>So there isn't a complete set of Logo memos online? I did get some help
from the MIT library services about AI memos. You could try there.

Not certainly so. There is complete set of Logo memos online, I think. But
they contain only mentions in relation CLOGO, like say AIM-307a does. And
there is no any memo with full description for CLOGO system like for
11LOGO.
However there should be such description. Because CLOGO was used as main
LOGO dialect at MIT before 11LOGO / LISP LOGO era. That is why I guess that
logo working papers may contain comprehensive CLOGO description. But MIT
logo related working papers much harder to find that AI MEMOs.


пн, 14 янв. 2019 г. в 11:32, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
So thanks again for the links. I'll try to do the logo working paper
request and get back with a result.

пн, 14 янв. 2019 г. в 11:55, Alexey Slyusar <aleksey.slyusar@gmail.com>:

**Comment by Alexey-Slyusar** [comment]
I've requested  "The LOGO Primer", LOGO Working Paper 19, MIT AI Lab (Jan. 1972) by Austin, H from MIT Institute Archives and Special Collections.
They've checked papers of Jean Bamberger (MC-0620) and double-checking the AI memos in case they have memo 19  but it wasn't cataloged. They tried a box of Seymour Papert's papers (MC-0693) also. But unfortunately, they did not find a copy of The LOGO Primer.
They kindly noted that they continually receive materials from faculty, offices, labs, centers, and departments, and it is good idea to not hesitate to contact they in the future in case they receive a copy. So there is still a hope to obtain the document.

**Comment by Alexey-Slyusar** [comment]
Thanks to the Lars's hint, following up on this in #1385, we've made clear issue related to creation the new Logo directory and new CLOGO user initiation. So now there is full information I need to finish the document.
The main problem that still remains regarding this issue is lack of time. It takes some more work than just add `FHACK` description to the document, because beside careful reading and fixing numerous typo there is another work. As we've already discussed with Lars, due to the fact that the document is relatively sizeable, it is better to convert it from plain text to INFO format for better navigation.

**Comment by larsbrinkhoff** [comment]
Thank you.  Please take all the time you need.  There's no hurry.

**Comment by Alexey-Slyusar** [comment]
Hi Lars, 
Could you please point me Is there any documentation about `INFO` format on ITS?
There are great GNU tools intended to work with 'texinfo' but I suspect that they are not applicable to produce ITS `INFO` format.

Thank you.

**Comment by larsbrinkhoff** [comment]
I haven't seen any documentation on ITS' INFO format, but I'll take a
look around.

However, from what I see the format is closely similar to GNU info.
I have loaded ITS INFO files into GNU info without any problems.
Maybe the other way doesn't work.

**Comment by larsbrinkhoff** [comment]
I did look, but I didn't find any documentation for the INFO format.

I suggest looking at other INFO files to get a feeling for the format.
From a quick look:

- Files are divided into sections called nodes.
- Nodes are separated by ^_ characters.
- The first line in a node should read "File: FOO  Node: BAR,  UP: BAZ,  Next: QUUX
- A node can have a menu starting with the text "* Menu:"
- Menu items should have the text "* Foobar::  Text"

**Comment by Alexey-Slyusar** [comment]
Lars, thank you for suggestion. This is helpful and It seems that your description covers most cases of ITS INFO formatting.
I am examining `RMAIL` info file on ITS using EMACS and besides `^_` I see `^_^L` things. 
`Mathlab Memo #3` "An Introduction to ITS for the MACSYMA User" describes  `^_` as "control-backarrow" and can be reproduced by typing `^?`
I'm going to try things in ITS EMACS to have proper feedback how a document looks like on ITS.

**Comment by larsbrinkhoff** [comment]
Yes, I see there are ^L characters too.  I don't know if they are
significant or not.  ITS source files are usually divided into pages
with ^L so maybe it's best to follow that convention.

No, ^_ and ^? are different characters.  To insert a ^_ type ^Q first.
To type ^_ I press Control, Shift, and - but your keyboard is probably
different.

**Comment by atsampson** [comment]
One minor incompatibility is that GNU info matches node names case-sensitively, where ITS info is case-insensitive -- so, for example, the menu in DECUUO 8 doesn't work. (This would be pretty easy to fix automatically...)
